### PR TITLE
Feat isc dhcp

### PR DESCRIPTION
**Pull request by bastelfreak**
this adds a new API endpoint, /dhcp, to add static ip/mac adresses

### CODE REVIEW

File: marmoset/dhcp/dhcp_config.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
+from .isc_dhcp_ldap_config import ISCDhcpLdapConfig
+from marmoset import validation
+import socket
+
+
+class DhcpConfig:
+    def __init__(self, mac, ip_address, gateway=None, networkmask=None):
+        self.mac = mac
```
Comment (bastelfreak) [inline marmoset/dhcp/dhcp_config.py:22]: is_mac() here?

File: marmoset/dhcp/dhcp_config.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+        if validation.is_cidr(ip_address):
+            self.ip_address = validation.get_ip_from_cidr(ip_address)
+
+            if self.networkmask is None:
+                self.networkmask = validation.get_nm_from_cidr(ip_address)
+
+            if self.gateway is None:
+                self.gateway = validation.get_gw_from_cidr(ip_address)
+        else:
+            self.ip_address = ip_address
```
Comment (bastelfreak) [inline marmoset/dhcp/dhcp_config.py:39]: is_ipv4() here?

File: marmoset/webserver/dhcp.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
+from flask import request, make_response
+from flask.ext.restful import reqparse, Resource, url_for, abort
+from werkzeug.exceptions import NotFound
+from .. import dhcp
```
Comment (bastelfreak) [inline marmoset/webserver/dhcp.py]: from marmoset?

File: marmoset/webserver/dhcp.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+def build_parameters(additional_statements):
+    global parser
+
+    parser = reqparse.RequestParser()
+    parser.add_argument('mac', type=str, required=True)
+    parser.add_argument('ip_address', type=str, required=True)
+    parser.add_argument('gateway', type=str, required=False, default=None)
+    parser.add_argument('networkmask', type=str, required=False, default=None)
+    parser.add_argument('dhcp_hostname', type=str, required=True)
```
Comment (bastelfreak) [inline marmoset/webserver/dhcp.py]: can we use this as an internal variable without access from the outside?

File: marmoset/validation.py [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+
+
+def get_nm_from_cidr(cidr):
+    data = get_cidr(cidr)
+    return data['nm']
+
+
+def get_gw_from_cidr(cidr):
+    data = get_cidr(cidr)
+    return data['gw']
```
Comment (bastelfreak) [inline marmoset/validation.py:56]: this file is so awesome <3

File: marmoset/__init__.py [hunk 6]
```diff
@@ -1,4 +1,4 @@
-from . import config, cli
+from . import config, cli, validation
```
Comment (bastelfreak) [inline marmoset/__init__.py]: can we use from marmoset here?

File: marmoset/__init__.py [hunk 7]
```diff
@@ -1,6 +1,9 @@
-from . import config, cli
+from marmoset import config
+from marmoset import cli
+from marmoset import validation
 
-def run(config_file = None):
+
+def run(config_file=None):
     cfg = config.load(config_file)
```
Comment (bastelfreak) [inline marmoset/__init__.py]: what happens if config_file is None?

File: marmoset/dhcp/dhcp_config.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+        self.dhcp_hostname = socket.getfqdn()
```
Comment (bastelfreak) [inline marmoset/dhcp/dhcp_config.py]: How much load will this generate if we have many requests? what about setting the FQDN in the config file and do a fallback to socket.getfqdn() is nothing is set?

File: marmoset/dhcp/dhcp_config.py [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+
+        self.mac = None
+        self.ip_address = None
+        self.gateway = None
+        self.networkmask = None
+        self.dhcp_hostname = None
+
+        self.set_settings(True, mac, ip_address, gateway, networkmask)
+
+    def set_settings(self, set_not_required_if_none=False, mac=None, ip_address=None, gateway=None, networkmask=None):
```
Comment (bastelfreak) [inline marmoset/dhcp/dhcp_config.py]: `set_not_required_if_none`

watman

Comment (sebastianrakel) [inline marmoset/dhcp/dhcp_config.py]: it was late, i need a parameter name and tadaaaaa! set_not_required_if_none

File: marmoset/dhcp/dhcp_config.py [hunk 10]
```diff
@@ -0,0 +1,68 @@
+from .isc_dhcp_ldap_config import ISCDhcpLdapConfig
+from marmoset import validation
+import socket
+
+
+class DhcpConfig:
+    def __init__(self, mac, ip_address, gateway=None, networkmask=None):
+        self.additional_statements = {}
```
Comment (bastelfreak) [inline marmoset/dhcp/dhcp_config.py:10]: this adds all other supplied params to the attribute additional_statements?

please have a look at https://github.com/virtapi/marmoset/issues/7#issuecomment-175262281

File: marmoset/__init__.py (hunk 7 above)
Comment (sebastianrakel) [inline marmoset/__init__.py:7]: it is only if you use a other configfile than default
later in the code the config file path will automated be set by `path.join(path.dirname(__file__), '../marmoset.conf')` if no configfile parameter was provided

File: marmoset/__init__.py [hunk 11]
```diff
@@ -1,6 +1,9 @@
-from . import config, cli
+from marmoset import config
+from marmoset import cli
+from marmoset import validation
 
-def run(config_file = None):
+
+def run(config_file=None):
```
Comment (bastelfreak) [inline marmoset/__init__.py]: what happens if config_file is None?

File: marmoset/dhcp/dhcp_config.py (hunk 10 above)
Comment (sebastianrakel) [inline marmoset/dhcp/dhcp_config.py:10]: this initialize only a dictionary for additonal_statements... here is nothing added

File: marmoset/__init__.py (hunk 11 above)
Comment (sebastianrakel) [inline marmoset/__init__.py]: look above...

File: marmoset/dhcp/dhcp_config.py (hunk 8 above)
Comment (sebastianrakel) [inline marmoset/dhcp/dhcp_config.py]: i will change that

Comment (sebastianrakel) [inline marmoset/dhcp/dhcp_config.py]: it uses now a config value, implemented in  b3c514a20325fd971b168c57d58e4f2edebf0086  and  8cd88d37b8ac621b709c85b5b3916f3678731991

File: marmoset/dhcp/dhcp_config.py (hunk 2 above)
Comment (sebastianrakel) [inline marmoset/dhcp/dhcp_config.py:39]: no validation in the constructor, it will be checked at the input of arguments

**Comment by sebastianrakel** [comment]
i think we can merge this now, but somebody should test it before

**Comment by bastelfreak** [comment]
here was rubbish

**Comment by bastelfreak** [comment]
tested locally, didn't found any issues.

**Pull request merged**